source/common/network/dns_resolver/dns_factory.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/cluster/v3/cluster.pb.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.pb.h"
+#include "envoy/extensions/filters/udp/dns_filter/v3alpha/dns_filter.pb.h"
+#include "envoy/extensions/network/dns_resolver/apple/v3/apple_dns_resolver.pb.h"
+#include "envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.pb.h"
+#include "envoy/network/dns.h"
+
+#include "source/common/config/utility.h"
+#include "source/common/runtime/runtime_features.h"
+
+namespace Envoy {
+namespace Network {
+
+const std::string cares_dns_resolver = "envoy.dns_resolver.cares";
+const std::string apple_dns_resolver = "envoy.dns_resolver.apple";
+const std::string dns_resolver_category = "envoy.network_dnsresolvers";
+
+class DnsResolverFactory : public Config::TypedFactory {
+public:
+  /**
+   * @returns a callback to create a DnsResolver.
+   */
+
+  virtual DnsResolverSharedPtr createDnsResolverCb(
+      Event::Dispatcher& dispatcher, const Api::Api& api,
+      const envoy::config::core::v3::TypedExtensionConfig& dns_resolver_config) PURE;
+
+  std::string category() const override { return dns_resolver_category; }
+};
+
+// Retrieve the DNS related configurations in the passed in @param config, and store the data into
+// @param typed_dns_resolver_config. The design behavior is:
+//
+// 1) If the config has typed_dns_resolver_config, and the corresponding DNS resolver factory
+//    is registered, copy it into typed_dns_resolver_config and use it.
+//
+// 2) Otherwise, fall back to the default behavior. i.e, check whether this is MacOS, and the
+//    run time flag: envoy.restart_features.use_apple_api_for_dns_lookups is enable. If it is,
+//    synthetic a AppleDnsResolverConfig object and pack it into typed_dns_resolver_config.
+
+// 3) If it is not MacOS, synthetic a CaresDnsResolverConfig object and pack it into
+// typed_dns_resolver_config.
+//    This can enable Envoy to use cares DNS library during DNS resolving process. The details are:
+// 3.1) if dns_resolution_config exists, copy it into CaresDnsResolverConfig,
+//      and pack CaresDnsResolverConfig into typed_dns_resolver_config.
+// 3.2) if dns_resolution_config doesn't exists, follow below behavior for backward compatibility:
+// 3.3) if config is DnsFilterConfig, pack an empty CaresDnsResolverConfig into
+//      typed_dns_resolver_config.
+// 3.4) For all others, copy config.use_tcp_for_dns_lookups into
+//      CaresDnsResolverConfig.dns_resolver_options.use_tcp_for_dns_lookups
+// 3.5) For ClusterConfig, one extra thing is to copy dns_resolvers into
+//      CaresDnsResolverConfig.resolvers,
+// 3.6) Then pack CaresDnsResolverConfig into typed_dns_resolver_config.
+
+template <class T>
```

Comment:
why is the function templated? Can you add some comments to the code about this design choice?

Also, I'd call this template type ConfigType rather than T. I don't think this is like a generic container like vector<T> that doesn't really care what it's containing.

Also I became curious whether T was referenced as a type name in the body but T is hard to search for.

Reply:
Currently, there are about four places(server.cc:boostrap.proto, cluster_factory_impl.cc:cluster.proto, dns_cache_impl.cc:dns_cache.proto, and dns_filter_resolver.h:dns_filter.proto ) have almost identical logic to construct the parameters before call function:  DispatcherImpl::createDnsResolver(resolvers, dns_resolver_options).

Those identical logics can easily be implemented by using a template function for code simplicity. Templates are for different configuration types.

Also, since we are changing the parameters into typed_dns_resolver_config, also need to implement logic, like how to deal with the new typed config, and how to deal with the #ifdef __APPLE__ in original code, thus have one function to hide all those logic can abstract out the details from the caller.

Yeah, changed T into ConfigType. It's not in function body, but in function parameter list.

Please take a look the latest implementation for this function and let me know whether it make sense.